src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/CollectTraceTests.cs

```diff
@@ -0,0 +1,155 @@
+        {
+            const string DefaultRuleName = "FunctionalTestRule";
+            const string EgressProvider = "TmpEgressProvider";
+            const string EventProviderName = "TestScenario";
+            const string StoppingEventName = "UniqueEvent";
+
+            string qualifiedEventName = (opcode == TraceEventOpcode.Info) ? StoppingEventName : FormattableString.Invariant($"{StoppingEventName}/{opcode}");
+
+            using TemporaryDirectory tempDirectory = new(_outputHelper);
+
+            Task ruleCompletedTask = null;
+
+            await ScenarioRunner.SingleTarget(_outputHelper,
+                _httpClientFactory,
+                DiagnosticPortConnectionMode.Listen,
+                TestAppScenarios.TraceEvents.Name,
+                appValidate: async (appRunner, apiClient) =>
+                {
+                    await appRunner.SendCommandAsync(TestAppScenarios.TraceEvents.Commands.EmitUniqueEvent);
+                    await ruleCompletedTask;
+                    await appRunner.SendCommandAsync(TestAppScenarios.TraceEvents.Commands.ShutdownScenario);
+                },
+                configureTool: runner =>
+                {
+                    runner.ConfigurationFromEnvironment.AddFileSystemEgress(EgressProvider, tempDirectory.FullName);
+                    runner.ConfigurationFromEnvironment.CreateCollectionRule(DefaultRuleName)
+                        .SetStartupTrigger()
+                        .AddCollectTraceAction(
+                            new EventPipeProvider[] {
+                                new EventPipeProvider()
+                                {
+                                    Name = EventProviderName,
+                                    Keywords = "-1"
+                                }
+                            },
+                            EgressProvider, options =>
+                            {
+                                options.Duration = duration ?? TimeSpan.Parse(ActionOptionsConstants.Duration_MaxValue);
+                                options.StoppingEvent = new TraceEventFilter()
+                                {
+                                    ProviderName = EventProviderName,
+                                    EventName = qualifiedEventName,
+                                    PayloadFilter = payloadFilter
+                                };
+                            });
+
+                    ruleCompletedTask = runner.WaitForCollectionRuleCompleteAsync(DefaultRuleName);
+                });
+
+            string[] files = Directory.GetFiles(tempDirectory.FullName, "*.nettrace", SearchOption.TopDirectoryOnly);
+            string traceFile = Assert.Single(files);
+            await ValidateNettraceFile(traceFile);
+        }
+
+        private async Task ValidateNettraceFile(string filePath)
+        {
+            byte[] expectedMagicToken = Encoding.UTF8.GetBytes("Nettrace");
```

Comment:
I think it would be worthwhile to use EventPipeEventSource to validate that the expected event is in the trace. Have `ValidateNettraceFile` return a `Task<bool>`, update `StopOnEventTestCore` to have a `bool expectEvevent` parameter, and it assert whether the event existing or not is expected. Let's not rely on TaskCanceledException due to the collection rule timeout occurring since that is ambiguous as to why the collection rule didn't finish in time.